### PR TITLE
[`pycodestyle`] Add blank line(s) rules (`E301`, `E302`, `E303`, `E304`, `E305`, `E306`)

### PR DESCRIPTION
**Pull request by hoel-bagard**
## Summary

This PR is part of https://github.com/charliermarsh/ruff/issues/2402, it adds the `E301`, `E302`, `E303`, `E304`, `E305`, `E306` error rules along with their fixes.

A first attempt at implementing `E305` using physical lines was done [here](https://github.com/charliermarsh/ruff/pull/4635), however since some operations are expensive when using physical lines, this implementation uses logical lines.
 
To be able to use `NonLogicalNewline` to detect blank lines, I have removed the check explicitly skipping them. I have modified the existing rules so that their behavior remains unchanged.

## Test Plan
The test fixture uses [the one from pycodestyle](https://github.com/PyCQA/pycodestyle/blob/main/testsuite/E30.py
) as is, but the notes in the file about which line should generate which error do not fully match the rules nor the implementation. I therefore did my best to manually check that what I did matches the rules/implementation in a way that makes sense.

For example, the rule [E306](https://www.flake8rules.com/rules/E306.html)'s example is not detected by pycodestyle, but this [false positive](https://github.com/PyCQA/pycodestyle/issues/1011) is.

### CODE REVIEW

File: crates/ruff/src/rules/pycodestyle/rules/logical_lines/blank_lines.rs [hunk 1]
```diff
@@ -0,0 +1,483 @@
...
+    tracked_vars: &mut BlankLinesTrackingVars,
+    indent_level: usize,
+    locator: &Locator,
+    stylist: &Stylist,
+    context: &mut LogicalLinesContext,
+) {
+    if line.is_empty() {
+        tracked_vars.blank_lines += 1;
+        tracked_vars.blank_characters += u32::try_from(line.text().len())
+            .expect("The number of blank characters should be relatively small");
```
Comment (MichaReiser) [inline crates/ruff/src/rules/pycodestyle/rules/logical_lines/blank_lines.rs:324]: Nit: `len` gives you the number of bytes and not the number of blank characters. The two are different for non-ascii characters. E.g. emojis are commonly represented as 4 UTF-8 bytes, but they only form 1-2 characters.

**Review by MichaReiser — COMMENTED** [review]
Thanks for working on this. I haven't reviewed the changes in full details yet because I first want to discuss the overall direction. 

I think using tokens to detect the blank lines is preferred over the physical lines. However, I would refrain from changing the definition of `LogicalLine`s to include empty lines too because it would then contradict [Python's official definition of logical lines](https://docs.python.org/3/reference/lexical_analysis.html#logical-lines) (logical lines cannot be empty), which would be rather confusing. 

Is my understanding correct, that the rule only tests for blank lines before and after functions, methods, and classes? 

If so, then I think it might be *easiest* to rewrite the rule as an AST based rule that runs for every top level function (you'll need to pass the next statement too, to avoid duplicates), and classes. It can use the `Locator` and then count the leading/trailing new lines.  

If the rules instead test for proper spacing between all statements, then I think it's best if we would build out a new `TokenVisitor` infrastructure that allows to runs some logic for every token in the program. Both `LogicalLine`s and your new rule could implement the `TokenVisitor` trait that has a single `visit_token(&mut self, token: &Tok, range: TextRange)` method. The visitor can track its own state internally and can trigger its rules when reaching a certain token (end of a logical line, end of a blank line). I would prefer this over changing the semantics of logical line because it still visits the tokens only once and provides us with a more extensible solution that could potentially even allow the other physical lines rule to work on tokens, rather than using the `UniversalNewlinesIterator`. @charliermarsh what do you think, would you have time to build such infrastructure?

File: crates/ruff/src/rules/pycodestyle/rules/logical_lines/blank_lines.rs [hunk 2]
```diff
@@ -0,0 +1,483 @@
...
+    }
+
+    for (token_idx, token) in line.tokens().iter().enumerate() {
+        // E301
+        if token.kind() == TokenKind::Def
+            && tracked_vars.is_in_class
+            && tracked_vars.blank_lines == 0
+            && prev_line
+                .and_then(|prev_line| prev_line.tokens_trimmed().first())
+                .map_or(false, |token| token.kind() != TokenKind::Class)
```
[the remark this answers is not in the document]
Comment (hoel-bagard) [reply crates/ruff/src/rules/pycodestyle/rules/logical_lines/blank_lines.rs:343]: Are you talking about a case like this ?

```python
class A:

    def b():
        pass
```

File: crates/ruff/src/rules/pycodestyle/rules/logical_lines/blank_lines.rs (hunk 1 above)
Comment (hoel-bagard) [reply crates/ruff/src/rules/pycodestyle/rules/logical_lines/blank_lines.rs]: Fixed in 47063d4b7536e5624cf96b6cc150a62ad54a417f

File: crates/ruff/src/rules/pycodestyle/rules/logical_lines/blank_lines.rs [hunk 3]
```diff
@@ -0,0 +1,483 @@
...
+        {
+            let mut diagnostic = Diagnostic::new(BlankLineBetweenMethods, token.range());
+            #[allow(deprecated)]
+            diagnostic.set_fix(Fix::unspecified(Edit::insertion(
+                stylist.line_ending().as_str().to_string(),
+                locator.line_start(token.range().start()),
+            )));
```
[the remark this answers is not in the document]
Comment (hoel-bagard) [reply crates/ruff/src/rules/pycodestyle/rules/logical_lines/blank_lines.rs]: Fixed in 2b4e95954dd8f47317eaf7c3e46ca125bb133ffd

File: crates/ruff/src/rules/pycodestyle/rules/logical_lines/blank_lines.rs [hunk 4]
```diff
@@ -0,0 +1,483 @@
...
+/// PEP 8 recommends the use of blank lines as follows:
+/// - Two blank lines are expected between functions and classes
+/// - One blank line is expected between methods of a class.
+///
+/// ## Example
+/// ```python
+/// def func1():
+///     pass
+/// def func2():
+///     pass
```
[the remark this answers is not in the document]
Comment (hoel-bagard) [reply crates/ruff/src/rules/pycodestyle/rules/logical_lines/blank_lines.rs:101]: My understanding is that some rules test for too few blank lines ([E301](https://www.flake8rules.com/rules/E301.html), [E302](https://www.flake8rules.com/rules/E302.html), [E305](https://www.flake8rules.com/rules/E305.html) and [E306](https://www.flake8rules.com/rules/E306.html)) and others for extraneous blank lines ( [E303](https://www.flake8rules.com/rules/E303.html), [E304](https://www.flake8rules.com/rules/E304.html))

I'm going to fix the titles/descriptions accordingly.

File: crates/ruff/src/rules/pycodestyle/rules/logical_lines/blank_lines.rs [hunk 5]
```diff
@@ -0,0 +1,483 @@
...
+pub struct BlankLineBetweenMethods;
+
+impl AlwaysAutofixableViolation for BlankLineBetweenMethods {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Expected 1 blank line, found 0")
+    }
+
+    fn autofix_title(&self) -> String {
+        "Remove extraneous blank line(s)".to_string()
```
[the remark this answers is not in the document]
Comment (hoel-bagard) [reply crates/ruff/src/rules/pycodestyle/rules/logical_lines/blank_lines.rs]: Fixed in [2fc94f8](https://github.com/charliermarsh/ruff/pull/4694/commits/2fc94f80fcaed9bc599842b91172c8a322671699)

**Comment by hoel-bagard** [comment]
That's essentially what I did (when the logical line is empty I just increment a counter and return), so I think that would work.
I'm tracking quite a few variables, so that might need to be refined.

**Comment by charliermarsh** [comment]
I looked into merging this but it doesn't seem to properly handle nested definitions -- this snippet is clean under pycodestyle, but yields violations here:

```py
class C:
    def f():
        pass


def f():
    def f():
        pass
```

Yields:

```
foo.py:2:5: E306 [*] Expected 1 blank line before a nested definition, found 0
foo.py:7:5: E302 [*] Expected 2 blank lines, found 0
foo.py:7:5: E306 [*] Expected 1 blank line before a nested definition, found 0
```

**Comment by hoel-bagard** [comment]
@charliermarsh Yes, the rules and examples of pycodestyles do not match the actual implementation.

For example, the rule [E306](https://www.flake8rules.com/rules/E306.html)'s example is not detected by pycodestyle, but this [false positive](https://github.com/PyCQA/pycodestyle/issues/1011) is.

**Comment by hoel-bagard** [comment]
I haven't had time to look into tacking the number of blank lines in the logical lines builder yet, but I'll try to do it once I have time.

**Comment by hoel-bagard** [comment]
I've moved the tracking of blank lines into the logical lines builder, so blank lines are no longer emitted as logical lines.
However, I'm sure how to track the number of blank characters there. I used to have `line.text().chars().count()`, but I don't think I can access the actual text from within the builder, so instead I used the tokens.

**Comment by hoel-bagard** [comment]
I started reviewing the output on the pycodestyle fixture, and found a major issue with this PR. Pycodestyle's E3 rules take into account comments. For example:
```python
a = 1




# a



# a
class A:
    pass
```

Pycodestyle output:
```
example.py:6:1: E303 too many blank lines (4)
example.py:10:1: E303 too many blank lines (3)
example.py:11:1: E302 expected 2 blank lines, found 4
```

However, since comment only lines are not logical lines, I can't reproduce this output. Moreover, this causes the implementation of the autofix to mess up the code.

Without making empty lines/comment only lines into logical lines (or using physical lines), I don't think these rules can be implemented.
@charliermarsh Would you have a recommendation on how I could proceed ?

**Comment by MichaReiser** [comment]
> If the rules instead test for proper spacing between all statements, then I think it's best if we would build out a new TokenVisitor infrastructure that allows to runs some logic for every token in the program. Both LogicalLines and your new rule could implement the TokenVisitor trait that has a single visit_token(&mut self, token: &Tok, range: TextRange) method. The visitor can track its own state internally and can trigger its rules when reaching a certain token (end of a logical line, end of a blank line). I would prefer this over changing the semantics of logical line because it still visits the tokens only once and provides us with a more extensible solution that could potentially even allow the other physical lines rule to work on tokens, rather than using the UniversalNewlinesIterator. @charliermarsh what do you think, would you have time to build such infrastructure?

@charliermarsh what do you think?

**Comment by akx** [comment]
Could we move this forward somehow? I was about to start implementing these rules but luckily checked whether there was momentum on it already :)

**Comment by hoel-bagard** [comment]
I'll finish cleaning the fixture to make it easier to check that the implementation works (or not).

But without implementing something like the proposed TokenVisitor, it's difficult to progress.

**Comment by hoel-bagard** [comment]
@MichaReiser Did something like the `TokenVisitor` get implemented since you talked about it ? I would like to finish the PR, but I can't do it using logical lines, since some cases do not include any logical line and should still emit an error (see the example below).

```python
# comment




# comment
```

**Comment by MichaReiser** [comment]
> @MichaReiser Did something like the `TokenVisitor` get implemented since you talked about it ? I would like to finish the PR, but I can't do it using logical lines, since some cases do not include any logical line and should still emit an error (see the example below).
> 
> ```python
> # comment
> 
> 
> 
> 
> # comment
> ```

Not the way I described it but there is a set of token-based rules that operate only on the tokens/indexer. Would that meet your needs?

https://github.com/astral-sh/ruff/blob/b34278e0cd25a24b69f4786897ca80c19fc3d5f1/crates/ruff_linter/src/checkers/tokens.rs#L22-L197

**Comment by hoel-bagard** [comment]
I think so, thank you.
Is there something I should keep in mind while using the tokens ? For example, in my attempt using physical lines it ended up too slow to be usable.

**Comment by MichaReiser** [comment]
> I should keep in mind while using the tokens ? For example, in my attempt using physical lines it ended up too slow to be usable.

Try to perform as many operations as possible on the tokens directly. Only fall back to inspect the source code (or token content) if you must.

**Comment by hoel-bagard** [comment]
Closed in favor of https://github.com/astral-sh/ruff/pull/8720.

**Pull request closed without merging**